meson_options.txt
# These options do not correspond to a --enable/--disable-* option
# on the configure script command line.  If you add more, list them in
# scripts/meson-buildoptions.py's SKIP_OPTIONS constant too.

option('qemu_suffix', type : 'string', value: 'qemu',
       description: 'Suffix for QEMU data/modules/config directories (can be empty)')
option('docdir', type : 'string', value : 'share/doc',
       description: 'Base directory for documentation installation (can be empty)')
option('qemu_firmwarepath', type : 'array', value : ['share/qemu-firmware'],
       description: 'search PATH for firmware files')
option('pkgversion', type : 'string', value : '',
       description: 'use specified string as sub-version of the package')
option('smbd', type : 'string', value : '',
       description: 'Path to smbd for slirp networking')
option('iasl', type : 'string', value : '',
       description: 'Path to ACPI disassembler')
option('tls_priority', type : 'string', value : 'NORMAL',
       description: 'Default TLS protocol/cipher priority string')
option('default_devices', type : 'boolean', value : true,
       description: 'Include a default selection of devices in emulators')
option('audio_drv_list', type: 'array', value: ['default'],
       choices: ['alsa', 'coreaudio', 'default', 'dsound', 'jack', 'oss', 'pa', 'pipewire', 'sdl', 'sndio'],
       description: 'Set audio driver list')
option('block_drv_rw_whitelist', type : 'string', value : '',
       description: 'set block driver read-write whitelist (by default affects only QEMU, not tools like qemu-img)')
option('block_drv_ro_whitelist', type : 'string', value : '',
       description: 'set block driver read-only whitelist (by default affects only QEMU, not tools like qemu-img)')
option('interp_prefix', type : 'string', value : '/usr/gnemul/qemu-%M',
       description: 'where to find shared libraries etc., use %M for cpu name')
option('rtsig_map', type : 'string', value : 'NULL',
       description: 'default value of QEMU_RTSIG_MAP')
option('fuzzing_engine', type : 'string', value : '',
       description: 'fuzzing engine library for OSS-Fuzz')
option('trace_file', type: 'string', value: 'trace',
       description: 'Trace file prefix for simple backend')
option('coroutine_backend', type: 'combo',
       choices: ['ucontext', 'sigaltstack', 'windows', 'wasm', 'auto'],
       value: 'auto', description: 'coroutine backend to use')
option('gdb', type: 'string',
       description: 'Path to GDB')

# Everything else can be set via --enable/--disable-* option
# on the configure script command line.  After adding an option
# here make sure to run "make update-buildoptions".

option('tests', type: 'feature', value: 'auto',
       description: 'Build the test suite')
option('docs', type : 'feature', value : 'auto',
       description: 'Documentations build support')
option('fuzzing', type : 'boolean', value: false,
       description: 'build fuzzing targets')
option('gettext', type : 'feature', value : 'auto',
       description: 'Localization of the GTK+ user interface')
option('modules', type : 'feature', value : 'disabled',
       description: 'modules support (non Windows)')
option('module_upgrades', type : 'boolean', value : false,
       description: 'try to load modules from alternate paths for upgrades')
option('install_blobs', type : 'boolean', value : true,
       description: 'install provided firmware blobs')
option('sparse', type : 'feature', value : 'auto',
       description: 'sparse checker')
option('guest_agent', type : 'feature', value : 'auto',
       description: 'Build QEMU Guest Agent')
option('guest_agent_msi', type : 'feature', value : 'auto',
       description: 'Build MSI package for the QEMU Guest Agent')
option('tools', type : 'feature', value : 'auto',
       description: 'build support utilities that come with QEMU')
option('qga_vss', type : 'feature', value: 'auto',
       description: 'build QGA VSS support (broken with MinGW)')

option('malloc_trim', type : 'feature', value : 'auto',
       description: 'enable libc malloc_trim() for memory optimization')
option('malloc', type : 'combo', choices : ['system', 'tcmalloc', 'jemalloc'],
       value: 'system', description: 'choose memory allocator to use')

option('kvm', type: 'feature', value: 'auto',
       description: 'KVM acceleration support')
option('mshv', type: 'feature', value: 'auto',
       description: 'MSHV acceleration support')
option('whpx', type: 'feature', value: 'auto',
       description: 'WHPX acceleration support')
option('hvf', type: 'feature', value: 'auto',
       description: 'HVF acceleration support')
option('nitro', type: 'feature', value: 'auto',
       description: 'Nitro acceleration support')
option('nvmm', type: 'feature', value: 'auto',
       description: 'NVMM acceleration support')
option('xen', type: 'feature', value: 'auto',
       description: 'Xen backend support')
option('xen_pci_passthrough', type: 'feature', value: 'auto',
       description: 'Xen PCI passthrough support')
option('tcg', type: 'feature', value: 'enabled',
       description: 'TCG support')
option('plugins', type: 'boolean', value: false,
       description: 'TCG plugins via shared library loading')
option('debug_tcg', type: 'boolean', value: false,
       description: 'TCG debugging')
option('debug_remap', type: 'boolean', value: false,
       description: 'syscall buffer debugging support')
option('tcg_interpreter', type: 'boolean', value: false,
       description: 'TCG with bytecode interpreter (slow)')
option('safe_stack', type: 'boolean', value: false,
       description: 'SafeStack Stack Smash Protection (requires clang/llvm and coroutine backend ucontext)')
option('asan', type: 'boolean', value: false,
       description: 'enable address sanitizer')
option('ubsan', type: 'boolean', value: false,
       description: 'enable undefined behaviour sanitizer')
option('tsan', type: 'boolean', value: false,
       description: 'enable thread sanitizer')
option('stack_protector', type: 'feature', value: 'auto',
       description: 'compiler-provided stack protection')
option('cfi', type: 'boolean', value: false,
       description: 'Control-Flow Integrity (CFI)')
option('cfi_debug', type: 'boolean', value: false,
       description: 'Verbose errors in case of CFI violation')
option('hmp', type : 'feature', value : 'auto',
       description: 'HMP monitor support')
option('multiprocess', type: 'feature', value: 'auto',
       description: 'Out of process device emulation support')
option('relocatable', type : 'boolean', value : true,
       description: 'toggle relocatable install')
option('vfio_user_server', type: 'feature', value: 'disabled',
       description: 'vfio-user server support')
option('dbus_display', type: 'feature', value: 'auto',
       description: '-display dbus support')
option('qemu_vnc', type: 'feature', value: 'auto',
       description: 'standalone VNC server over D-Bus')
option('tpm', type : 'feature', value : 'auto',
       description: 'TPM support')
option('valgrind', type : 'feature', value: 'auto',
       description: 'valgrind debug support for coroutine stacks')
option('igvm', type: 'feature', value: 'auto',
       description: 'Independent Guest Virtual Machine (IGVM) file support')

# Do not enable it by default even for Mingw32, because it doesn't
# work on Wine.
option('membarrier', type: 'feature', value: 'disabled',
       description: 'membarrier system call (for Linux 4.14+ or Windows')

option('keyring', type: 'feature', value: 'auto',
       description: 'Linux keyring support')
option('libkeyutils', type: 'feature', value: 'auto',
       description: 'Linux keyutils support')

option('af_xdp', type : 'feature', value : 'auto',
       description: 'AF_XDP network backend support')
option('attr', type : 'feature', value : 'auto',
       description: 'attr/xattr support')
option('auth_pam', type : 'feature', value : 'auto',
       description: 'PAM access control')
option('brlapi', type : 'feature', value : 'auto',
       description: 'brlapi character device driver')
option('bzip2', type : 'feature', value : 'auto',
       description: 'bzip2 support for DMG images')
option('cap_ng', type : 'feature', value : 'auto',
       description: 'cap_ng support')
option('blkio', type : 'feature', value : 'auto',
       description: 'libblkio block device driver')
option('bpf', type : 'feature', value : 'auto',
        description: 'eBPF support')
option('cocoa', type : 'feature', value : 'auto',
       description: 'Cocoa user interface (macOS only)')
option('curl', type : 'feature', value : 'auto',
       description: 'CURL block device driver')
option('gio', type : 'feature', value : 'auto',
       description: 'use libgio for D-Bus support')
option('hv_balloon', type : 'feature', value : 'auto',
       description: 'hv-balloon driver (requires Glib 2.68+ GTree API)')
option('libdw', type : 'feature', value : 'auto',
       description: 'debuginfo support')
option('libiscsi', type : 'feature', value : 'auto',
       description: 'libiscsi userspace initiator')
option('libnfs', type : 'feature', value : 'auto',
       description: 'libnfs block device driver')
option('mpath', type : 'feature', value : 'auto',
       description: 'Multipath persistent reservation passthrough')
option('numa', type : 'feature', value : 'auto',
       description: 'libnuma support')
option('iconv', type : 'feature', value : 'auto',
       description: 'Font glyph conversion support')
option('curses', type : 'feature', value : 'auto',
       description: 'curses UI')
option('libcbor', type : 'feature', value : 'auto',
       description: 'libcbor support')
option('gnutls', type : 'feature', value : 'auto',
       description: 'GNUTLS cryptography support')
option('nettle', type : 'feature', value : 'auto',
       description: 'nettle cryptography support')
option('gcrypt', type : 'feature', value : 'auto',
       description: 'libgcrypt cryptography support')
option('crypto_afalg', type : 'feature', value : 'disabled',
       description: 'Linux AF_ALG crypto backend driver')
option('libdaxctl', type : 'feature', value : 'auto',
       description: 'libdaxctl support')
option('libpmem', type : 'feature', value : 'auto',
       description: 'libpmem support')
option('libssh', type : 'feature', value : 'auto',
       description: 'ssh block device support')
option('libudev', type : 'feature', value : 'auto',
       description: 'Use libudev to enumerate host devices')
option('libusb', type : 'feature', value : 'auto',
       description: 'libusb support for USB passthrough')
option('linux_aio', type : 'feature', value : 'auto',
       description: 'Linux AIO support')
option('linux_io_uring', type : 'feature', value : 'auto',
       description: 'Linux io_uring support')
option('lzfse', type : 'feature', value : 'auto',
       description: 'lzfse support for DMG images')
option('lzo', type : 'feature', value : 'auto',
       description: 'lzo compression support')
option('pvg', type: 'feature', value: 'auto',
       description: 'macOS paravirtualized graphics support')
option('rbd', type : 'feature', value : 'auto',
       description: 'Ceph block device driver')
option('opengl', type : 'feature', value : 'auto',
       description: 'OpenGL support')
option('rdma', type : 'feature', value : 'auto',
       description: 'Enable RDMA-based migration')
option('gtk', type : 'feature', value : 'auto',
       description: 'GTK+ user interface')
option('sdl', type : 'feature', value : 'auto',
       description: 'SDL user interface')
option('sdl_image', type : 'feature', value : 'auto',
       description: 'SDL Image support for icons')
option('seccomp', type : 'feature', value : 'auto',
       description: 'seccomp support')
option('smartcard', type : 'feature', value : 'auto',
       description: 'CA smartcard emulation support')
option('snappy', type : 'feature', value : 'auto',
       description: 'snappy compression support')
option('spice', type : 'feature', value : 'auto',
       description: 'Spice server support')
option('spice_protocol', type : 'feature', value : 'auto',
       description: 'Spice protocol support')
option('u2f', type : 'feature', value : 'auto',
       description: 'U2F emulation support')
option('canokey', type : 'feature', value : 'auto',
       description: 'CanoKey support')
option('usb_redir', type : 'feature', value : 'auto',
       description: 'libusbredir support')
option('l2tpv3', type : 'feature', value : 'auto',
       description: 'l2tpv3 network backend support')
option('netmap', type : 'feature', value : 'auto',
       description: 'netmap network backend support')
option('pixman', type : 'feature', value : 'auto',
       description: 'pixman support')
option('slirp', type: 'feature', value: 'auto',
       description: 'libslirp user mode network backend support')
option('passt', type: 'feature', value: 'auto',
       description: 'passt network backend support')
option('vde', type : 'feature', value : 'auto',
       description: 'vde network backend support')
option('vmnet', type : 'feature', value : 'auto',
       description: 'vmnet.framework network backend support')
option('virglrenderer', type : 'feature', value : 'auto',
       description: 'virgl rendering support')
option('rutabaga_gfx', type : 'feature', value : 'auto',
       description: 'rutabaga_gfx support')
option('png', type : 'feature', value : 'auto',
       description: 'PNG support with libpng')
option('vnc', type : 'feature', value : 'auto',
       description: 'VNC server')
option('vnc_jpeg', type : 'feature', value : 'auto',
       description: 'JPEG lossy compression for VNC server')
option('vnc_sasl', type : 'feature', value : 'auto',
       description: 'SASL authentication for VNC server')
option('vte', type : 'feature', value : 'auto',
       description: 'vte support for the gtk UI')
option('xkbcommon', type : 'feature', value : 'auto',
       description: 'xkbcommon support')
option('zstd', type : 'feature', value : 'auto',
       description: 'zstd compression support')
option('qpl', type : 'feature', value : 'auto',
       description: 'Query Processing Library support')
option('uadk', type : 'feature', value : 'auto',
       description: 'UADK Library support')
option('qatzip', type: 'feature', value: 'auto',
       description: 'QATzip compression support')
option('fuse', type: 'feature', value: 'auto',
       description: 'FUSE block device export')
option('fuse_lseek', type : 'feature', value : 'auto',
       description: 'SEEK_HOLE/SEEK_DATA support for FUSE exports')

option('trace_backends', type: 'array', value: ['log'],
       choices: ['dtrace', 'ftrace', 'log', 'nop', 'simple', 'syslog', 'ust'],
       description: 'Set available tracing backends')

option('alsa', type: 'feature', value: 'auto',
       description: 'ALSA sound support')
option('coreaudio', type: 'feature', value: 'auto',
       description: 'CoreAudio sound support')
option('dsound', type: 'feature', value: 'auto',
       description: 'DirectSound sound support')
option('jack', type: 'feature', value: 'auto',
       description: 'JACK sound support')
option('oss', type: 'feature', value: 'auto',
       description: 'OSS sound support')
option('pa', type: 'feature', value: 'auto',
       description: 'PulseAudio sound support')
option('pipewire', type: 'feature', value: 'auto',
       description: 'PipeWire sound support')
option('sndio', type: 'feature', value: 'auto',
       description: 'sndio sound support')

option('vhost_kernel', type: 'feature', value: 'auto',
       description: 'vhost kernel backend support')
option('vhost_net', type: 'feature', value: 'auto',
       description: 'vhost-net kernel acceleration support')
option('vhost_user', type: 'feature', value: 'auto',
       description: 'vhost-user backend support')
option('vhost_crypto', type: 'feature', value: 'auto',
       description: 'vhost-user crypto backend support')
option('vhost_vdpa', type: 'feature', value: 'auto',
       description: 'vhost-vdpa kernel backend support')
option('vhost_user_blk_server', type: 'feature', value: 'auto',
       description: 'build vhost-user-blk server')
option('virtfs', type: 'feature', value: 'auto',
       description: 'virtio-9p support')
option('libvduse', type: 'feature', value: 'auto',
       description: 'build VDUSE Library')
option('vduse_blk_export', type: 'feature', value: 'auto',
       description: 'VDUSE block export support')

option('capstone', type: 'feature', value: 'auto',
       description: 'Whether and how to find the capstone library')

ChangeLog
2006-10-01 18:08  bellard

	* pc-bios/: bios.bin, bios.diff: synced to Bochs BIOS - use 32 bit
	  pushf/popf in 32 bit PCI bios - moved some useful defines in
	  rombios.h

2006-10-01 15:03  pbrook

	* hw/arm_gic.c: ARM GIC bug.

2006-09-29 22:01  bellard

	* target-i386/helper.c: 32 bit RSP update fix (aka Open Solaris
	  x86_64 bug)

2006-09-28 21:52  pbrook

	* hw/pci.c: PCI IRC routing fix.

2006-09-27 23:31  bellard

	* target-i386/helper.c: fixed user mode emulation

2006-09-27 21:54  bellard

	* target-i386/: cpu.h, helper.c: SMM fix for x86_64

2006-09-27 21:52  bellard

	* target-i386/helper2.c: enabled PSE36 for x86_64 (fix for
	  OpenSolaris as guest)

2006-09-25 23:41  bellard

	* hw/vga.c: VBE 4 bit bank selection fix (Volker Rupper)

2006-09-25 23:34  bellard

	* configure: removed unused code

2006-09-25 23:33  bellard

	* monitor.c: fixed help info strings

2006-09-24 21:36  bellard

	* pc-bios/: bios.bin, bios.diff: updated

2006-09-24 21:31  bellard

	* hw/piix_pci.c: implement i440 instead of i450 ISA memory mappings
	  to be compatible with Bochs

2006-09-24 20:49  bellard

	* Changelog: update

2006-09-24 20:49  bellard

	* vl.h: moved PCI init to BIOS

2006-09-24 20:49  bellard

	* hw/piix_pci.c: moved PCI init to BIOS - added ISA memory mapping
	  registers and SMM support

2006-09-24 20:48  bellard

	* linux-user/main.c: added cpu_smm_update()

2006-09-24 20:48  bellard

	* hw/pc.c: moved MP table init to BIOS

2006-09-24 20:45  bellard

	* Makefile.target, hw/acpi-dsdt.hex, hw/acpi-dsdt.dsl: moved ACPI
	  table init to BIOS

2006-09-24 20:44  bellard

	* hw/acpi.c: moved ACPI table init to BIOS - preliminary SMM
	  support

2006-09-24 20:41  bellard

	* exec.c: added cpu_get_physical_page_desc()

2006-09-24 20:40  bellard

	* target-i386/cpu.h, target-i386/exec.h, target-i386/helper.c,
	  target-i386/helper2.c, target-i386/op.c, target-i386/translate.c,
	  cpu-all.h, cpu-exec.c, vl.c: added SMM support

2006-09-24 20:38  bellard

	* pc-bios/: bios.bin, bios.diff: moved PCI, MP and ACPI init to
	  bios

2006-09-24 19:01  pbrook

	* vl.h, hw/apb_pci.c, hw/grackle_pci.c, hw/pci.c, hw/piix_pci.c,
	  hw/prep_pci.c, hw/unin_pci.c, hw/versatile_pci.c: Implement sun4u
	  PCI IRQ routing.  Allow multiple PCI busses and PCI-PCI bridges.
	  Fix bugs in Versatile PCI implementation.

2006-09-24 02:16  pbrook

	* vl.h, hw/apb_pci.c, hw/grackle_pci.c, hw/pci.c, hw/piix_pci.c,
	  hw/prep_pci.c, hw/unin_pci.c, hw/versatile_pci.c: PCI shared IRQ
	  fix (original patch by andrzej zaborowski).

2006-09-23 19:40  pbrook

	* Makefile.target, vl.c, vl.h, hw/arm_gic.c, hw/arm_sysctl.c,
	  hw/pl080.c, hw/realview.c, hw/versatile_pci.c, hw/versatilepb.c:
	  Add ARM RealView Emulation Baseboard.

2006-09-21 23:46  bellard

	* hw/: vga.c, vga_int.h: VBE: 8 bit DACs + support for VBE BIOS IDs
	  (Volker Ruppert)

2006-09-21 22:34  pbrook

	* pc-bios/openbios-sparc32: Update sparc bios image (Blue Swirl).

2006-09-20 22:30  bellard

	* cpu-exec.c, hw/slavio_misc.c: sun4m halt support (Blue Swirl)

2006-09-20 22:28  bellard

	* qemu-doc.texi: update

2006-09-18 03:15  pbrook

	* Makefile.target, vl.h, hw/isa_mmio.c, hw/mips_r4k.c, hw/ppc.c,
	  hw/ppc_chrp.c: Merge common ISA access routines.

2006-09-17 05:20  pbrook

	* hw/esp.c: SPARC SCSI fixes.

2006-09-10 21:25  bellard

	* Makefile.target, vl.h, hw/cs4231.c, hw/sun4m.c: dummy cs4231
	  audio driver for sun4m (Blue Swirl)

2006-09-10 16:39  pbrook

	* vl.c: Avoid (repeatedly) trying to read stdin after it has
	  closed.

2006-09-10 01:35  pbrook

	* linux-user/syscall_defs.h: ARM 64-bit host fix.

2006-09-09 16:36  pbrook

	* target-arm/helper.c: Arm MMU fix (Justin Fletcher).

2006-09-09 14:17  bellard

	* hw/slavio_serial.c: allow host serial port usage (Blue Swirl)

2006-09-09 14:03  bellard

	* block-vvfat.c: vvfat fixes (Roger Lathrop)

2006-09-09 13:38  bellard

	* hw/slavio_serial.c: fix serial irq logic (Blue Swirl)

2006-09-09 13:35  bellard

	* hw/slavio_serial.c: added mouse protocol (Igor Kovalenko)

2006-09-09 13:31  bellard

	* hw/tcx.c: more correct display functions

2006-09-09 13:10  bellard

	* vl.c: two stop bits support

2006-09-03 21:48  bellard

	* vl.h, hw/pcnet.c, hw/sparc32_dma.c: better PCNET endianness
	  support

2006-09-03 19:10  bellard

	* qemu-doc.texi: typo

2006-09-03 19:09  bellard

	* target-i386/translate.c: multi byte nop support

2006-09-03 18:40  bellard

	* hw/pcnet.c: endianness fixes

2006-09-03 18:09  bellard

	* Makefile.target, vl.h, hw/esp.c, hw/iommu.c, hw/sparc32_dma.c,
	  hw/sun4m.c: Separate the DMA controllers - Convert ESP to new DMA
	  methods (Blue Swirl)

2006-09-03 18:07  bellard

	* hw/: lance.c, pcnet.c: initial sparc32 lance and pcnet merge

2006-09-03 16:10  bellard

	* qemu-doc.texi, vl.c: allow disabling of serial or parallel
	  devices (Stefan Weil)

2006-09-03 14:08  pbrook

	* block-raw.c, vl.c, vl.h: Flush IO requests before savevm
	  (original patch by Blue Swirl).

2006-08-29 06:52  pbrook

	* vl.h, hw/esp.c, hw/iommu.c, hw/lsi53c895a.c, hw/scsi-disk.c,
	  hw/sun4m.c, hw/usb-msd.c: SCSI TCQ support.

2006-08-26 20:00  bellard

	* hw/serial.c: 16 bit divider

2006-08-24 22:36  bellard

	* vnc.c: signed fixes - VNC console switch (initial patch by
	  Andrzej Zaborowski)

2006-08-24 21:53  bellard

	* block.c: fixed handling of relative filenames with -snapshot

2006-08-24 20:58  bellard

	* block-raw.c: win32 aio emulation fix

2006-08-23 23:40  bellard

	* block-raw.c, block.c: win32 compilation fixes

2006-08-23 23:14  bellard

	* block.c: fixed error handling

2006-08-21 22:28  bellard

	* qemu-doc.texi: update

2006-08-21 22:26  bellard

	* qemu-doc.texi: spelling fixes (Stefan Weil)

2006-08-19 18:56  bellard

	* qemu-doc.texi: removed outdated info

2006-08-19 16:27  bellard

	* sdl.c, keymaps/ja: Japanese keyboard fixes (kazu)

2006-08-19 14:38  bellard

	* Changelog: update

2006-08-19 14:37  bellard

	* tap-win32.c, vl.c, vl.h: win32 tap poll suppression (kazu)


include/hw/xen/interface/hvm/params.h
/* SPDX-License-Identifier: MIT */
/*
 * Copyright (c) 2007, Keir Fraser
 */

#ifndef __XEN_PUBLIC_HVM_PARAMS_H__
#define __XEN_PUBLIC_HVM_PARAMS_H__

#include "hvm_op.h"

/* These parameters are deprecated and their meaning is undefined. */
#if defined(__XEN__) || defined(__XEN_TOOLS__)

#define HVM_PARAM_PAE_ENABLED                4
#define HVM_PARAM_DM_DOMAIN                 13
#define HVM_PARAM_MEMORY_EVENT_CR0          20
#define HVM_PARAM_MEMORY_EVENT_CR3          21
#define HVM_PARAM_MEMORY_EVENT_CR4          22
#define HVM_PARAM_MEMORY_EVENT_INT3         23
#define HVM_PARAM_NESTEDHVM                 24
#define HVM_PARAM_MEMORY_EVENT_SINGLE_STEP  25
#define HVM_PARAM_BUFIOREQ_EVTCHN           26
#define HVM_PARAM_MEMORY_EVENT_MSR          30

#endif /* defined(__XEN__) || defined(__XEN_TOOLS__) */

/*
 * Parameter space for HVMOP_{set,get}_param.
 */

#define HVM_PARAM_CALLBACK_IRQ 0
#define HVM_PARAM_CALLBACK_IRQ_TYPE_MASK xen_mk_ullong(0xFF00000000000000)
/*
 * How should CPU0 event-channel notifications be delivered?
 *
 * If val == 0 then CPU0 event-channel notifications are not delivered.
 * If val != 0, val[63:56] encodes the type, as follows:
 */

#define HVM_PARAM_CALLBACK_TYPE_GSI      0
/*
 * val[55:0] is a delivery GSI.  GSI 0 cannot be used, as it aliases val == 0,
 * and disables all notifications.
 */

#define HVM_PARAM_CALLBACK_TYPE_PCI_INTX 1
/*
 * val[55:0] is a delivery PCI INTx line:
 * Domain = val[47:32], Bus = val[31:16] DevFn = val[15:8], IntX = val[1:0]
 */

#if defined(__i386__) || defined(__x86_64__)
#define HVM_PARAM_CALLBACK_TYPE_VECTOR   2
/*
 * val[7:0] is a vector number.  Check for XENFEAT_hvm_callback_vector to know
 * if this delivery method is available.
 */
#elif defined(__arm__) || defined(__aarch64__)
#define HVM_PARAM_CALLBACK_TYPE_PPI      2
/*
 * val[55:16] needs to be zero.
 * val[15:8] is interrupt flag of the PPI used by event-channel:
 *  bit 8: the PPI is edge(1) or level(0) triggered
 *  bit 9: the PPI is active low(1) or high(0)
 * val[7:0] is a PPI number used by event-channel.
 * This is only used by ARM/ARM64 and masking/eoi the interrupt associated to
 * the notification is handled by the interrupt controller.
 */
#define HVM_PARAM_CALLBACK_TYPE_PPI_FLAG_MASK      0xFF00
#define HVM_PARAM_CALLBACK_TYPE_PPI_FLAG_LOW_LEVEL 2
#endif

/*
 * These are not used by Xen. They are here for convenience of HVM-guest
 * xenbus implementations.
 */
#define HVM_PARAM_STORE_PFN    1
#define HVM_PARAM_STORE_EVTCHN 2

#define HVM_PARAM_IOREQ_PFN    5

#define HVM_PARAM_BUFIOREQ_PFN 6

#if defined(__i386__) || defined(__x86_64__)

/*
 * Viridian enlightenments
 *
 * (See http://download.microsoft.com/download/A/B/4/AB43A34E-BDD0-4FA6-BDEF-79EEF16E880B/Hypervisor%20Top%20Level%20Functional%20Specification%20v4.0.docx)
 *
 * To expose viridian enlightenments to the guest set this parameter
 * to the desired feature mask. The base feature set must be present
 * in any valid feature mask.
 */
#define HVM_PARAM_VIRIDIAN     9

/* Base+Freq viridian feature sets:
 *
 * - Hypercall MSRs (HV_X64_MSR_GUEST_OS_ID and HV_X64_MSR_HYPERCALL)
 * - APIC access MSRs (HV_X64_MSR_EOI, HV_X64_MSR_ICR and HV_X64_MSR_TPR)
 * - Virtual Processor index MSR (HV_X64_MSR_VP_INDEX)
 * - Timer frequency MSRs (HV_X64_MSR_TSC_FREQUENCY and
 *   HV_X64_MSR_APIC_FREQUENCY)
 */
#define _HVMPV_base_freq 0
#define HVMPV_base_freq  (1 << _HVMPV_base_freq)

/* Feature set modifications */

/* Disable timer frequency MSRs (HV_X64_MSR_TSC_FREQUENCY and
 * HV_X64_MSR_APIC_FREQUENCY).
 * This modification restores the viridian feature set to the
 * original 'base' set exposed in releases prior to Xen 4.4.
 */
#define _HVMPV_no_freq 1
#define HVMPV_no_freq  (1 << _HVMPV_no_freq)

/* Enable Partition Time Reference Counter (HV_X64_MSR_TIME_REF_COUNT) */
#define _HVMPV_time_ref_count 2
#define HVMPV_time_ref_count  (1 << _HVMPV_time_ref_count)

/* Enable Reference TSC Page (HV_X64_MSR_REFERENCE_TSC) */
#define _HVMPV_reference_tsc 3
#define HVMPV_reference_tsc  (1 << _HVMPV_reference_tsc)

/* Use Hypercall for remote TLB flush */
#define _HVMPV_hcall_remote_tlb_flush 4
#define HVMPV_hcall_remote_tlb_flush (1 << _HVMPV_hcall_remote_tlb_flush)

/* Use APIC assist */
#define _HVMPV_apic_assist 5
#define HVMPV_apic_assist (1 << _HVMPV_apic_assist)

/* Enable crash MSRs */
#define _HVMPV_crash_ctl 6
#define HVMPV_crash_ctl (1 << _HVMPV_crash_ctl)

/* Enable SYNIC MSRs */
#define _HVMPV_synic 7
#define HVMPV_synic (1 << _HVMPV_synic)

/* Enable STIMER MSRs */
#define _HVMPV_stimer 8
#define HVMPV_stimer (1 << _HVMPV_stimer)

/* Use Synthetic Cluster IPI Hypercall */
#define _HVMPV_hcall_ipi 9
#define HVMPV_hcall_ipi (1 << _HVMPV_hcall_ipi)

/* Enable ExProcessorMasks */
#define _HVMPV_ex_processor_masks 10
#define HVMPV_ex_processor_masks (1 << _HVMPV_ex_processor_masks)

/* Allow more than 64 VPs */
#define _HVMPV_no_vp_limit 11
#define HVMPV_no_vp_limit (1 << _HVMPV_no_vp_limit)

/* Enable vCPU hotplug */
#define _HVMPV_cpu_hotplug 12
#define HVMPV_cpu_hotplug (1 << _HVMPV_cpu_hotplug)

#define HVMPV_feature_mask \
        (HVMPV_base_freq | \
         HVMPV_no_freq | \
         HVMPV_time_ref_count | \
         HVMPV_reference_tsc | \
         HVMPV_hcall_remote_tlb_flush | \
         HVMPV_apic_assist | \
         HVMPV_crash_ctl | \
         HVMPV_synic | \
         HVMPV_stimer | \
         HVMPV_hcall_ipi | \
         HVMPV_ex_processor_masks | \
         HVMPV_no_vp_limit | \
         HVMPV_cpu_hotplug)

#endif

/*
 * Set mode for virtual timers (currently x86 only):
 *  delay_for_missed_ticks (default):
 *   Do not advance a vcpu's time beyond the correct delivery time for
 *   interrupts that have been missed due to preemption. Deliver missed
 *   interrupts when the vcpu is rescheduled and advance the vcpu's virtual
 *   time stepwise for each one.
 *  no_delay_for_missed_ticks:
 *   As above, missed interrupts are delivered, but guest time always tracks
 *   wallclock (i.e., real) time while doing so.
 *  no_missed_ticks_pending:
 *   No missed interrupts are held pending. Instead, to ensure ticks are
 *   delivered at some non-zero rate, if we detect missed ticks then the
 *   internal tick alarm is not disabled if the VCPU is preempted during the
 *   next tick period.
 *  one_missed_tick_pending:
 *   Missed interrupts are collapsed together and delivered as one 'late tick'.
 *   Guest time always tracks wallclock (i.e., real) time.
 */
#define HVM_PARAM_TIMER_MODE   10
#define HVMPTM_delay_for_missed_ticks    0
#define HVMPTM_no_delay_for_missed_ticks 1
#define HVMPTM_no_missed_ticks_pending   2
#define HVMPTM_one_missed_tick_pending   3

/* Boolean: Enable virtual HPET (high-precision event timer)? (x86-only) */
#define HVM_PARAM_HPET_ENABLED 11

/* Identity-map page directory used by Intel EPT when CR0.PG=0. */
#define HVM_PARAM_IDENT_PT     12

/* ACPI S state: currently support S0 and S3 on x86. */
#define HVM_PARAM_ACPI_S_STATE 14

/* TSS used on Intel when CR0.PE=0. */
#define HVM_PARAM_VM86_TSS     15

/* Boolean: Enable aligning all periodic vpts to reduce interrupts */
#define HVM_PARAM_VPT_ALIGN    16

/* Console debug shared memory ring and event channel */
#define HVM_PARAM_CONSOLE_PFN    17
#define HVM_PARAM_CONSOLE_EVTCHN 18

/*
 * Select location of ACPI PM1a and TMR control blocks. Currently two locations
 * are supported, specified by version 0 or 1 in this parameter:
 *   - 0: default, use the old addresses
 *        PM1A_EVT == 0x1f40; PM1A_CNT == 0x1f44; PM_TMR == 0x1f48
 *   - 1: use the new default qemu addresses
 *        PM1A_EVT == 0xb000; PM1A_CNT == 0xb004; PM_TMR == 0xb008
 * You can find these address definitions in <hvm/ioreq.h>
 */
#define HVM_PARAM_ACPI_IOPORTS_LOCATION 19

/* Params for the mem event rings */
#define HVM_PARAM_PAGING_RING_PFN   27
#define HVM_PARAM_MONITOR_RING_PFN  28
#define HVM_PARAM_SHARING_RING_PFN  29

/* SHUTDOWN_* action in case of a triple fault */
#define HVM_PARAM_TRIPLE_FAULT_REASON 31

#define HVM_PARAM_IOREQ_SERVER_PFN 32
#define HVM_PARAM_NR_IOREQ_SERVER_PAGES 33

/* Location of the VM Generation ID in guest physical address space. */
#define HVM_PARAM_VM_GENERATION_ID_ADDR 34

/*
 * Get mode for altp2m:
 *  disabled: altp2m not active (default)
 *  mixed: allow access to all altp2m ops for both in-guest and external tools
 *  external: allow access to external privileged tools only
 *  limited: guest only has limited access (ie. control VMFUNC and #VE)
 */
#define HVM_PARAM_ALTP2M       35
#define XEN_ALTP2M_disabled      0
#define XEN_ALTP2M_mixed         1
#define XEN_ALTP2M_external      2
#define XEN_ALTP2M_limited       3

/*
 * Size of the x87 FPU FIP/FDP registers that the hypervisor needs to
 * save/restore.  This is a workaround for a hardware limitation that
 * does not allow the full FIP/FDP and FCS/FDS to be restored.
 *
 * Valid values are:
 *
 * 8: save/restore 64-bit FIP/FDP and clear FCS/FDS (default if CPU
 *    has FPCSDS feature).
 *
 * 4: save/restore 32-bit FIP/FDP, FCS/FDS, and clear upper 32-bits of
 *    FIP/FDP.
 *
 * 0: allow hypervisor to choose based on the value of FIP/FDP
 *    (default if CPU does not have FPCSDS).
 *
 * If FPCSDS (bit 13 in CPUID leaf 0x7, subleaf 0x0) is set, the CPU
 * never saves FCS/FDS and this parameter should be left at the
 * default of 8.
 */
#define HVM_PARAM_X87_FIP_WIDTH 36

/*
 * TSS (and its size) used on Intel when CR0.PE=0. The address occupies
 * the low 32 bits, while the size is in the high 32 ones.
 */
#define HVM_PARAM_VM86_TSS_SIZED 37

/* Enable MCA capabilities. */
#define HVM_PARAM_MCA_CAP 38
#define XEN_HVM_MCA_CAP_LMCE   (xen_mk_ullong(1) << 0)
#define XEN_HVM_MCA_CAP_MASK   XEN_HVM_MCA_CAP_LMCE

#define HVM_NR_PARAMS 39

#endif /* __XEN_PUBLIC_HVM_PARAMS_H__ */
